sbt-plugin/src/main/scala/sbt/ch/epfl/scala/ProfilingSbtPlugin.scala

```diff
@@ -26,81 +26,138 @@ object ProfilingSbtPlugin extends AutoPlugin {
-    TimeUnit.MILLISECONDS.toSeconds(if (time == null) 0 else time.toLong)
-  }
-
   val buildSettings: Seq[Def.Setting[_]] = Nil
   val projectSettings: Seq[Def.Setting[_]] = List(
-    BuildKeys.profilingWarmupCompiler in Compile := Def.taskDyn {
-      val logger = Keys.streams.value.log
-      val currentCompiler = Keys.compilers.value
-      val warmupDuration = BuildKeys.profilingWarmupDuration.value.toLong
-      val currentExecutionTime = getWarmupTime(currentCompiler)
-
-      if (currentExecutionTime == 0)
-        logger.info(s"Warming up compiler ($currentExecutionTime out of $warmupDuration)...")
-
-      val scopedKey: sbt.ScopedKey[_] = Keys.resolvedScoped.value
-      val compileKey = Keys.compile in scopedKey.scope
-      val cleanKey = Keys.clean in scopedKey.scope
-      val timeExecution = timeCompileExecution(compileKey, currentCompiler)
-      def recurse = Def.taskDyn {
-        val t = getWarmupTime(currentCompiler)
-        loop(t)
-      }
+    BuildKeys.profilingWarmupCompiler :=
+      BuildDefaults.profilingWarmupCompiler.tag(WarmupTag, Tags.Compile).value
+  )
 
-      def loop(currentDuration: Long): Def.Initialize[Task[Unit]] = {
-        if (currentDuration < warmupDuration) Def.task {
-          if (currentDuration != 0)
-            logger.info(s"Warming up compiler ($currentDuration out of $warmupDuration)...")
-          StableDef.sequential(compileKey, timeExecution, cleanKey, recurse).value
-        } else StableDef.task(logger.success(s"The compiler is warmed up (${warmupDuration}s)."))
+  object BuildDefaults {
+    import sbt.{Command, State}
+    import sbt.complete.Parser
+
+    val profilingWarmupCompiler: Def.Initialize[Task[Unit]] = Def.task {
+      // Meh, we don't care about the resulting state, we'll throw it away.
+      def runCommandAndRemaining(command: String): State => State = { st: State =>
+        @annotation.tailrec
+        def runCommand(command: String, state: State): State = {
+          val nextState = Parser.parse(command, state.combinedParser) match {
+            case Right(cmd) => cmd()
+            case Left(msg) => sys.error(s"Invalid programmatic input:\n$msg")
+          }
+          nextState.remainingCommands.toList match {
+            case Nil => nextState
+            case head :: tail => runCommand(head, nextState.copy(remainingCommands = tail))
+          }
+        }
+        runCommand(command, st.copy(remainingCommands = Nil))
+          .copy(remainingCommands = st.remainingCommands)
       }
 
-      loop(getWarmupTime(currentCompiler))
-    }.value,
-    BuildKeys.profilingWarmupDuration := BuildDefaults.profilingWarmupDuration.value
-  )
+      // The fact that we cannot call `name` on a recently created Command is pretty ugly.
+      val commandName = profilingWarmupCommand.asInstanceOf[sbt.SimpleCommand].name
+      val tweakedState = Keys.state.value.put(BuildKeys.currentProject, Keys.thisProjectRef.value)
```

Comment:
We need to do this because sbt does not resolve `thisProjectRef` and `extracted.currentRef` to the right project that invoked this task, even though the state inside the task should have the right value.